src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs

```diff
@@ -994,6 +1166,280 @@ public void CanBindValueTypeOptions()
+        }
+
+        [Fact]
+        public void CanBindSemiImmutableClass_WithInitProperties()
+        {
+            var dic = new Dictionary<string, string>
+            {
+                {"Length", "42"},
+                {"Color", "Green"},
+                {"Thickness", "1.23"},
+            };
+            var configurationBuilder = new ConfigurationBuilder();
+            configurationBuilder.AddInMemoryCollection(dic);
+            var config = configurationBuilder.Build();
+
+            var options = config.Get<SemiImmutableClassWithInit>();
+            Assert.Equal(42, options.Length);
+            Assert.Equal("Green", options.Color);
+            Assert.Equal(1.23m, options.Thickness);
+        }
+
+        [Fact]
+        public void CanBindRecordOptions()
+        {
+            var dic = new Dictionary<string, string>
+            {
+                {"Length", "42"},
+                {"Color", "Green"},
+            };
+            var configurationBuilder = new ConfigurationBuilder();
+            configurationBuilder.AddInMemoryCollection(dic);
+            var config = configurationBuilder.Build();
+
+            var options = config.Get<RecordTypeOptions>();
+            Assert.Equal(42, options.Length);
+            Assert.Equal("Green", options.Color);
+        }
+
+        [Fact]
+        public void CanBindStructOptions()
+        {
+            var dic = new Dictionary<string, string>
+            {
+                {"Length", "42"},
+                {"Color", "Green"},
+            };
+            var configurationBuilder = new ConfigurationBuilder();
+            configurationBuilder.AddInMemoryCollection(dic);
+            var config = configurationBuilder.Build();
+
+            var options = config.Get<StructTypeOptions>();
+            Assert.Equal(42, options.Length);
+            Assert.Equal("Green", options.Color);
+        }
+
+        [Fact]
+        public void CanBindReadonlyStructOptions()
+        {
+            var dic = new Dictionary<string, string>
+            {
+                {"Length", "42"},
+                {"Color", "Green"},
+            };
+            var configurationBuilder = new ConfigurationBuilder();
+            configurationBuilder.AddInMemoryCollection(dic);
+            var config = configurationBuilder.Build();
+
+            var options = config.Get<ReadonlyStructTypeOptions>();
+            Assert.Equal(42, options.Length);
+            Assert.Equal("Green", options.Color);
+        }
+
+        [Fact]
+        public void CanBindRecordStructOptions()
+        {
+            var dic = new Dictionary<string, string>
+            {
+                {"Length", "42"},
+                {"Color", "Green"},
+            };
+            var configurationBuilder = new ConfigurationBuilder();
+            configurationBuilder.AddInMemoryCollection(dic);
+            var config = configurationBuilder.Build();
+
+            var options = config.Get<RecordStructTypeOptions>();
```

Comment:
If we're sticking with both calling the constructor and setting properties, shouldn't we avoid doing both? Right now, this test effectively calls `new RecordStructTypeOptions("Green", 42) { Color = "Green", Length = 42 }` which is redudant.

We should at least track the properties we've already set via constructor parameters and not set them again. As for,

> But on the other hand, a user might want a best-effort approach, e.g. I have a constructor that "I" use, and I'd like binding to bind to whatever properties it can. 

I still think we should just reject types with parameter names that don't match any property like S.T.J. That's different than rejecting types with property names that don't match parameter names. I think it's better to catch subtle bugs where the parameter and the property names don't match up early.

I don't think it's super important, and I think it would be better off to just not to support both passing in constructor arguments and setting parameters at least initially. But if you think it's useful, filling in *additional* properties if they don't match any parameters doesn't seem so bad.

Reply:
Fair point. While I personally think it's useful, I do see your point that setting properties twice is redundant.

> I still think we should just reject types with parameter names that don't match any property like S.T.J
Is this comment specific to records, or generally? I added some tests for records where the primary constructor had a property name that matched an existing property:
```csharp
public record struct RecordStructTypeOptions2(string Color, int Length)
{
    private readonly string _color = "";

    public string Color
    {
        get => _color;
        init => _color = "the color is " + value;
    }
}
```
This results in a compilation error for the parameter of:
` error CS8907: Parameter 'Color' is unread. Did you forget to use it to initialize the property with that name?`

Reply:
Is the overhead of tracking which properties have been set worth it here?

For these kinds of objects, setting the properties should be super fast right? Are we concerned about:

1. The setter doing something different than what the ctor parameter would do?
2. The overhead of getting the value from the `IConfiguration` twice?

Reply:
In my opinion, _potentially_ setting properties twice (the parameter in the constructor may not set a property), is OK. Yes, it's redundancy, but as you say, it should be super fast. However, it's a fair point re. the concern of reading from config twice.

Reply:
I spoke with @halter73 offline, and I agree with his concern here. If I was to write this code by hand (or in a source generator), I wouldn't pass the value into both the constructor as well as set the property. So I think it makes sense to remember which values were passed into the constructor, and skip those properties.

Reply:
I also think we should also throw if there isn't any property matching the parameter to align with S.T.J:

```csharp
using System.Text.Json;
using System.Text.Json.Serialization;

public class Test
{
    [JsonConstructor]
    public Test(string a, string b, string c)
    {
        A = a;
        B = b;
    }

    public string A { get; }
    public string B { get; }
}

// System.InvalidOperationException:
//   Each parameter in the deserialization constructor on type 'Test'
//   must bind to an object property or field on deserialization.
JsonSerializer.Deserialize<Test>(@"{
    ""A"": ""a"",
    ""B"": ""b""
    ""C"": ""c""
}");
```

Reply:
This is now implemented

Reply:
Interestingly, while playing around with this example some more, I came to the conclusion that the exception is either misleading, **or** there is a bug in STJ. I added a **field** named C, and set it in the constructor, but still got the same exception. I then made each property and field, and _still_ got the same exception!

```csharp
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ConsoleApp1
{
    internal class Program
    {
        static void Main(string[] args)
        {
// System.InvalidOperationException:
//   Each parameter in the deserialization constructor on type 'Test'
//   must bind to an object property or field on deserialization.
            JsonSerializer.Deserialize<Test>(@"{
    ""A"": ""a"",
    ""B"": ""b"",
    ""C"": ""c""
}");
        }
    }
    
    public class Test
    {
        [JsonConstructor]
        public Test(string a, string b, string c)
        {
            A = a;
            B = b;
            C = c;
        }

        public string A {get;}
        public string B {get; }
        public string C;
    }

}


```

Reply:
I've debugged this test in S.T.J, and have discovered that the exception above is thrown if the `ParameterCache` count is not the same as the parameter count of the construct.  **But**, the `ParameterCache` property says this:

![image](https://user-images.githubusercontent.com/263416/170812689-1972eba2-d1cd-4e26-90bd-508315e65af4.png)

Which doesn't match the text of the exception:

`Each parameter in the deserialization constructor on type 'Test' must bind to an object property or field on deserialization.`

I'm not sure if it's the code or the exception message that's at fault.

Reply:
@davidfowl replied to that bug report and said that config is required to include fields. I created a very small PR to small changes to the wording of the exception. But it begs the question here: should we have the same behaviour as S.T.J regarding matching constructor parameters to fields/properties. We can decide on 'fields by default' after, as it may be moot depending on the outcome of this question.

Reply:
Just to summarise this thread: S.T.J. has an _option_ to consider fields. I added that too, but we've decided that we won't consider fields, just properties. This behaviour is now implemented.